qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,864 @@
+
+        fargs = _get_expression_args(self._envelope, self.parameters)
+        return Waveform(samples=self._envelope_lam(*fargs), name=self.name)
+
+    def validate_parameters(self) -> None:
+        """Validate parameters.
+
+        Raises:
+            PulseError: If the parameters passed are not valid.
+        """
+        if self.is_parameterized():
+            return
+
+        if self._constraints is not None:
+            fargs = _get_expression_args(self._constraints, self.parameters)
+            if not bool(self._constraints_lam(*fargs)):
+                param_repr = ", ".join(f"{p}={v}" for p, v in self.parameters.items())
+                const_repr = str(self._constraints)
+                raise PulseError(
+                    f"Assigned parameters {param_repr} violate following constraint: {const_repr}."
+                )
+
+        if self.limit_amplitude:
+            if self._valid_amp_conditions is not None:
+                fargs = _get_expression_args(self._valid_amp_conditions, self.parameters)
+                check_full_waveform = not bool(self._valid_amp_conditions_lam(*fargs))
+            else:
+                check_full_waveform = True
+
+            if check_full_waveform:
+                # Check full waveform only when the condition is satisified or
+                # evaluation condition is not provided.
+                # This operation is slower due to overhead of 'get_waveform'.
+                if not _is_amplitude_valid(self):
+                    param_repr = ", ".join(f"{p}={v}" for p, v in self.parameters.items())
+                    raise PulseError(
+                        f"Maximum pulse amplitude norm exceeds 1.0 with parameters {param_repr}."
+                        "This can be overruled by setting Pulse.limit_amplitude."
+                    )
+
+    def is_parameterized(self) -> bool:
+        """Return True iff the instruction is parameterized."""
+        return any(isinstance(val, ParameterExpression) for val in self.parameters.values())
+
+    @property
+    def parameters(self) -> Dict[str, Any]:
+        params = {"duration": self.duration}
+        params.update(self._params)
+        return params
+
+    def __eq__(self, other: "SymbolicPulse") -> bool:
+
+        if not isinstance(other, SymbolicPulse):
+            return NotImplemented
+
+        if self._pulse_type != other._pulse_type:
+            return False
+
+        if self._envelope != other._envelope:
+            return False
+
+        if self.parameters != other.parameters:
+            return False
+
+        return True
+
+    def __hash__(self) -> int:
+        # TODO hashing symbolic and parametric pulse is no longer correct behavior.
+        #  Remove the hash if possible -- but this might be used in user code.
+        #  We need polite deprecation to remove this.
+        return hash((self._pulse_type, self.duration, *tuple(self._params.items())))
```

Comment:
In https://github.com/Qiskit/qiskit-terra/pull/7821#discussion_r895153218, you suggest deprecating `__hash__` which seems okay to me. From what you say there, it seems like the main use cases for the hash are Qobj serialization and caching of amplitude validation via get_waveform. Both of those cases require that the pulse is no longer parametrized. Could we raise here if the pulse is parametrized? That would prevent the case of the parameter values being reassigned after the hash has been computed.

Also, I don't understand why you don't include the expressions in the hash.

Reply:
Adding parameter check is perfect idea. Done in https://github.com/Qiskit/qiskit-terra/pull/7821/commits/cda7336eab3937e5d7cd942908267cee2afda0f0.

Ignoring the expression in hashing is because it is targeting qobj generation for now, where it only cares the pulse type name and assigned parameters. This must be changed after we switch to QPY model. But I'm still open to adding expression if there is already reasonable reason for it.

Reply:
Well, we have the following for `_is_amplitude_valid` right now:

https://github.com/Qiskit/qiskit-terra/blob/ef35cd9da61921bc41c5aa929eb1d04b1844ef3b/qiskit/pulse/library/symbolic_pulses.py#L78-L95

`lru_cache()` effectively makes a tuple out of the function's arguments and stores the function's return value in a dictionary with that tuple as the key and then uses that key to look up and return the value on future calls without re-evaluating the function. That process of using the function arguments as a dictionary key requires hashing the arguments. So `SymbolicPulse`s are also hashed for this amplitude check since that is the input to the function. It is a very niche case to consider that `_pulse_type` could be the same while the `envelope` expressions are different, but that was the case I was thinking of. For `_is_amplitude_valid` it is the envelope (and parameters) that needs to be uniquely matched to the return value.

An alternative to including the envelope in the `SymbolicPulse` hash would be to refactor `_is_amplitude_valid` to be a function of the envelope and parameters directly. Then `lru_cache` would hash them directly. The complication with doing this is that `_is_amplitude_valid` wants the lambdified envelope and looking that up without the `SymbolicPulse` class is a little ugly (I think it is `SymbolicPulse.__dict__["_envelope_lam"].lambda_funcs[hash(envelope)]` unless you use the trick described [here](https://stackoverflow.com/a/21629855)).

Reply:
Make sense. I added envelope in the hash in https://github.com/Qiskit/qiskit-terra/pull/7821/commits/15e9f7c9e68038401ba6c197da6c42b37f6a5ddc